foglamp/FogLAMP
# FOGL-2012/2010 - Add API support for configuring notifications

To test this PR, we need to mock the Notification server. I have created a [mock Python Notification server](https://gist.github.com/amarendra-dianomic/366d82db7c7b973a7fbcfc3808529c04) at port 8118 which can be used, otherwise one is at liberty to use other mock server provided the mock server is able to service desired rest calls. You will need to download this mock Notification server (testnotification.py) to a folder.

Steps to test:
1. Start Foglamp.
2. In a separate terminal, start the mock Notification Server (via python3 testnotification.py). Please note that everytime Foglamp server is started, you will need to start/restart the mock Notification server.
3. Hit curl -X GET http://localhost:8081/foglamp/service | jq to verify the newly added service registry entry for Notification service.
4. Add Notification: Hit curl -X POST http://localhost:8081/foglamp/notification -d '{"name": "Test Notification", "description":"Test Notification", "rule": "threshold", "channel": "email", "notification_type": "one shot", "enabled": false}'
5. Get all Notifications: Hit curl -X GET http://localhost:8081/foglamp/notification | jq to get the list.
6. Get Notification detail: Hit curl -X GET http://localhost:8081/foglamp/notification/Test%20Notification | jq
7. Add another Notification with new rule_config and delivery_config: Hit curl -X POST http://localhost:8081/foglamp/notification -d '{"name": "Test Notification2", "description":"Test Notification", "rule": "threshold", "channel": "email", "notification_type": "one shot", "enabled": false, "rule_config":{"window": {"value":"100"}}, "delivery_config": {"server": {"value":"pop"}} }'
8. Get all Notifications: Hit curl -X GET http://localhost:8081/foglamp/notification | jq to get the list.
9. Update a notification: Hit  curl -X PUT http://localhost:8081/foglamp/notification/Test%20Notification -d '{"rule_config": {"window": {"default": "60","value": "60","description": "The window to perform rule evaluation over in minutes","type": "integer"},"tolerance": {"default": "4","value": "4","description": "The accepted tolerance","type": "integer"},"trigger": {"default": "80","value": "80","description": "Temparature threshold value","type": "integer"},"asset": {"default": "temperature","value": "temperature","description": "The asset the notification is defined against","type": "string"},"plugin": {"default": "threshold","value": "threshold","description": "The accepted non tolerance","type": "string"}}}'
10. Get Notification detail: Hit curl -X GET http://localhost:8081/foglamp/notification/Test%20Notification | jq to verify the new values of trigger and plugin description of the rule_config.
11. Delete Notification: Hit curl -X DELETE http://localhost:8081/foglamp/notification/Test%20Notification | jq.
12. Get all Notifications: Hit curl -X GET http://localhost:8081/foglamp/notification | jq to get the list and verify.
13. Hit curl -X GET http://localhost:8081/foglamp/notification/plugin | jq to see the list of available rule plugins and delivery plugins. 

All the above are positive use cases. We are not supposed to add a rule or a delivery channel which is not present in the get_plugins list. Similarly, we are also not supposed to add/delete/modify any config component of rule_config and/or delivery_config. We may change the value only.
